includes/rest/class-server.php

```diff
@@ -170,4 +172,64 @@ public static function request_parameter_order( $order, $request ) {
 			'defaults',
 		);
 	}
+
+	/**
+	 * Filters the REST API response to properly handle the ActivityPub error formatting.
+	 *
+	 * @see https://codeberg.org/fediverse/fep/src/branch/main/fep/c180/fep-c180.md
+	 *
+	 * @param WP_HTTP_Response $response Result to send to the client. Usually a `WP_REST_Response`.
+	 * @param WP_REST_Server   $server   Server instance.
+	 * @param WP_REST_Request  $request  Request used to generate the response.
+	 *
+	 * @return WP_HTTP_Response The filtered response.
+	 */
+	public static function filter_output( $response, $server, $request ) {
+		$route = $request->get_route();
+
+		// Check if it is an activitypub request and exclude webfinger and nodeinfo endpoints.
+		if ( ! \str_starts_with( $route, '/' . ACTIVITYPUB_REST_NAMESPACE ) ) {
+			return $response;
+		}
+
+		// Only alter responses that return an error status code.
+		if ( $response->get_status() < 300 ) {
+			return $response;
+		}
+
+		$data  = $response->get_data();
+		$error = array(
+			'type'     => 'about:blank',
+			'title'    => isset( $data['code'] ) ? $data['code'] : '',
+			'detail'   => isset( $data['message'] ) ? $data['message'] : '',
+			'status'   => $response->get_status(),
+
+			/*
+			 * Provides the unstructured error data.
+			 *
+			 * @see https://nodeinfo.diaspora.software/schema.html#metadata.
+			 */
+			'metadata' => $data,
+		);
+
+		switch ( $data['code'] ) {
+			case 'activitypub_signature':
+			case 'activitypub_signature_verification':
+				$error['title'] = 'Invalid Signature';
+				break;
+			case 'activitypub_comment_not_found':
+			case 'activitypub_local_only_comment':
+			case 'activitypub_post_not_found':
+			case 'activitypub_user_not_found':
+				$error['type']   = 'https://w3id.org/fep/c180#object-does-not-exist';
+				$error['status'] = 400;
+				$error['title']  = 'Object does not exist';
```

Comment:
If I understand the spec correctly, this is about activities referencing a non-existing object, which is/might be different from what these `*_not_found` error codes refer to?

> The activity refers to an object in one of its properties, such as object, target, or an addressing property, but the object does not exist.

Reply:
> Note that this type is distinct from an endpoint returning a 404 Not Found status code for a GET request for an object that does not exist, or for posting to an endpoint that does not exist.

🤔

This makes it even more complicated :(